File: 101. Symmetric Tree—Breadth-first Search/101. Symmetric Tree—Breadth-first Search.java
//1. 树是空的，true。

//2. 判断这棵树的左子树和右子树是否对称。

//3. 两棵子树的对称条件：根节点相等，左子树的左子树和右子树的右子树对称，左子树的右子树和右子树的左子树对称：

 

//       2            2
//      / \          / \
//     3   4        4   3
//    / \ / \      / \ / \
//   5  6 7  8    8  7 6  5
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        else {
            return isSymmetric(root.left, root.right);
        }
    }
    private boolean isSymmetric(TreeNode left, TreeNode right) {
        if (left == null && right == null) {
            return true;
        }
        else if (left == null || right == null) {
            return false;
        }
        return left.val == right.val && isSymmetric(left.left, right.right) && isSymmetric(left.right, right.left);
    }
}

File: 101. Symmetric Tree—Breadth-first Search/101. Symmetric Tree—Breadth-first Search.cpp
#--------------多年未见cpp写的深搜---------------------
class Solution {
    bool sameTree(TreeNode* a, TreeNode* b) {
        if (a == NULL && b == NULL) return true;
        if (a == NULL || b == NULL) return false;
        if (a->val != b->val) return false;
        return sameTree(a->left, b->right) && sameTree(a->right, b->left);
    }
public:
    bool isSymmetric(TreeNode* root) {
        if (root == NULL) return true;
        return sameTree(root->left, root->right);
    }
};
#----------------------二---------------------
#说明: 
#如果你可以运用递归和迭代两种方法解决这个问题，会很加分。

#思路：
#递归终止条件：：两棵树同时遍历到空（true），仅一棵树遍历到空（false），两棵树遍历到的节点值不同（false）。
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    bool rec(TreeNode* leftR,TreeNode* rightR){
        if(leftR == NULL && rightR == NULL)
            return true;
        if(leftR == NULL || rightR == NULL)
            return false;
        if(leftR->val != rightR->val)
            return false;
        return (rec(leftR->left,rightR->right) && rec(leftR->right,rightR->left));
    }
public:
    bool isSymmetric(TreeNode* root) {
        if(root == NULL)
            return true;
        return rec(root->left,root->right);      
    }
};
